transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -462,46 +611,80 @@ public ChannelHandler replace(String oldName, String newName, ChannelHandler new
 
     private ChannelHandler replace(
             final AbstractChannelHandlerContext ctx, String newName, ChannelHandler newHandler) {
-
         assert ctx != head && ctx != tail;
 
-        Future<?> future;
+        final AbstractChannelHandlerContext newCtx;
+        final EventExecutor executor;
+        final boolean inEventLoop;
         synchronized (this) {
+            checkMultiplicity(newHandler);
+
             if (newName == null) {
                 newName = ctx.name();
             } else if (!ctx.name().equals(newName)) {
                 newName = filterName(newName, newHandler);
             }
 
-            final AbstractChannelHandlerContext newCtx =
-                    new DefaultChannelHandlerContext(this, ctx.invoker, newName, newHandler);
-
-            if (!isExecuteLater(newCtx)) {
+            newCtx = new DefaultChannelHandlerContext(this, ctx.invoker, newName, newHandler);
+            executor = executorSafe(ctx.invoker);
+            if (executor == null) {
                 replace0(ctx, newCtx);
+                callHandlerCallbackLater(newCtx, true);
+                callHandlerCallbackLater(ctx, false);
                 return ctx.handler();
+            }
+            inEventLoop = executor.inEventLoop();
+            if (inEventLoop) {
+                replace0(ctx, newCtx);
+            }
+        }
+        if (inEventLoop) {
+            // Invoke newHandler.handlerAdded() first (i.e. before oldHandler.handlerRemoved() is invoked)
+            // because callHandlerRemoved() will trigger inboundBufferUpdated() or flush() on newHandler and those
+            // event handlers must be called after handlerAdded().
+            callHandlerAdded0(newCtx);
+
+            EventExecutor eventExecutor = ctx.executor();
```

Comment:
Is this a different object than the `executor` above? (`executor = executorSafe(ctx.invoker)`) I may be missing something, but they look like a same object, because this method returns when `executorSafe()` returns `null`.